src/storage-controller/src/lib.rs

```diff
@@ -2999,35 +2997,18 @@ where
                         // from the finalization register, unfortunately... but the next run
                         // of `finalize_shards` should notice the upper has advanced and tidy
                         // up.
+                        let persist_client = persist_client.clone();
                         mz_ore::task::spawn(|| format!("finalize_shard({shard_id})"), async move {
-                            let result =
-                                tokio::time::timeout(
-                                    Duration::from_secs(15 * 60),
-
-                                    write_handle.append(
-                                        Vec::<(
-                                            (mz_storage_types::sources::SourceData, ()),
-                                            T,
-                                            Diff,
-                                        )>::new(),
-                                        write_handle.upper().clone(),
-                                        Antichain::new(),
-                                    )
-                                ).await;
-
-                            // Rather than error, just leave this shard as one to finalize later.
-                            match result {
-                                Err(_) => {
-                                    warn!("timed out while trying to finalize shard {shard_id}");
-                                }
-                                Ok(Err(usage)) => {
-                                    error!("invalid usage while finalizing shard {shard_id}: {usage:?}")
-                                }
-                                Ok(Ok(Err(mismatch))) => {
-                                    warn!("unable to advance the upper of shard {shard_id} to the empty antichain: {mismatch:?}")
-                                }
-                                Ok(Ok(Ok(()))) => {}
-                            };
+                            let result = persist_client
+                                .finalize_shard::<SourceData, (), T, Diff>(
```

Comment:
I haven't loaded the storage controller side of this up again, so this might be naive, but don't we need to keep the append? (I'm thinking that we might we get stuck never finalizing it because the upper never gets advanced to the empty antichain?)

Reply:
No, you're right - this is sketchy. I've restored the call.

It's still not ideal that we're using the upper being `[]` as a proxy for being tombstoned, since that's not guaranteed by either the new or old code. Perhaps we should expose a new is_tombstone method somewhere?

Happy to try and either figure that out here or save it for a followup, as you prefer.

Reply:
> It's still not ideal that we're using the upper being [] as a proxy for being tombstoned, since that's not guaranteed by either the new or old code. Perhaps we should expose a new is_tombstone method somewhere?

Hmmm good point. Let's hash this out in our 1:1 today

Reply:
For the record: we decided to add the method to make this unambiguous as part of this PR. (This is done!)